File: educational-core/src/com/jetbrains/edu/coursecreator/actions/placeholder/CCDeleteAnswerPlaceholder.java
package com.jetbrains.edu.coursecreator.actions.placeholder;

import com.intellij.openapi.actionSystem.Presentation;
import com.intellij.openapi.project.Project;
import com.jetbrains.edu.learning.EduState;
import com.jetbrains.edu.learning.EduUtils;
import com.jetbrains.edu.learning.courseFormat.AnswerPlaceholder;
import com.jetbrains.edu.learning.courseFormat.TaskFile;
import com.jetbrains.edu.learning.messages.EduCoreBundle;
import org.jetbrains.annotations.NonNls;
import org.jetbrains.annotations.NotNull;

public class CCDeleteAnswerPlaceholder extends CCAnswerPlaceholderAction {

  @NonNls
  public static final String ACTION_ID = "Educational.Educator.DeleteAnswerPlaceholder";

  public CCDeleteAnswerPlaceholder() {
    super();
  }

  @Override
  protected void performAnswerPlaceholderAction(@NotNull Project project, @NotNull EduState state) {
    deletePlaceholder(project, state);
  }

  private static void deletePlaceholder(@NotNull Project project, @NotNull EduState state) {
    TaskFile taskFile = state.getTaskFile();
    AnswerPlaceholder answerPlaceholder = state.getAnswerPlaceholder();
    if (answerPlaceholder == null) {
      throw new IllegalStateException("Delete Placeholder action called, but no placeholder found");
    }

    EduUtils.runUndoableAction(project, EduCoreBundle.message("action.Educational.Educator.DeleteAnswerPlaceholder.text.full"),
                               new CCAddAnswerPlaceholder.AddAction(project, answerPlaceholder, taskFile, state.getEditor()) {
                                 @Override
                                 public void undo() {
                                   super.redo();
                                 }

                                 @Override
                                 public void redo() {
                                   super.undo();
                                 }
                               });
  }

  private static boolean canDeletePlaceholder(@NotNull EduState state) {
    if (state.getEditor().getSelectionModel().hasSelection()) {
      return false;
    }
    return state.getAnswerPlaceholder() != null;
  }

  @Override
  protected void updatePresentation(@NotNull EduState eduState, @NotNull Presentation presentation) {
    presentation.setEnabledAndVisible(canDeletePlaceholder(eduState));
  }
}

File: educational-core/src/com/jetbrains/edu/coursecreator/actions/placeholder/CCDeleteAnswerPlaceholder.kt
package com.jetbrains.edu.coursecreator.actions.placeholder

import com.intellij.openapi.actionSystem.Presentation
import com.intellij.openapi.project.Project
import com.jetbrains.edu.learning.EduState
import com.jetbrains.edu.learning.actions.runUndoableAction
import com.jetbrains.edu.learning.messages.EduCoreBundle.message
import org.jetbrains.annotations.VisibleForTesting

class CCDeleteAnswerPlaceholder : CCAnswerPlaceholderAction() {
  override fun performAnswerPlaceholderAction(project: Project, state: EduState) {
    val taskFile = state.taskFile
    val answerPlaceholder = state.answerPlaceholder
                            ?: throw IllegalStateException("Delete Placeholder action called, but no placeholder found")
    runUndoableAction(project, message("action.Educational.Educator.DeleteAnswerPlaceholder.text.full"),
      object : CCAddAnswerPlaceholder.AddAction(project, answerPlaceholder, taskFile, state.editor) {
        override fun undo() {
          super.redo()
        }

        override fun redo() {
          super.undo()
        }
      })
  }

  override fun updatePresentation(eduState: EduState, presentation: Presentation) {
    if (eduState.editor.selectionModel.hasSelection()) {
      presentation.isEnabledAndVisible = false
    }
    else {
      presentation.isEnabledAndVisible = eduState.answerPlaceholder != null
    }
  }

  companion object {
    @VisibleForTesting
    const val ACTION_ID: String = "Educational.Educator.DeleteAnswerPlaceholder"
  }
}
